seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/serialize/ElasticsearchRowSerializer.java

```diff
@@ -45,15 +47,63 @@ public class ElasticsearchRowSerializer implements SeaTunnelRowSerializer {
     private final IndexSerializer indexSerializer;
 
     private final IndexTypeSerializer indexTypeSerializer;
+    private final Function<SeaTunnelRow, String> keyExtractor;
 
     public ElasticsearchRowSerializer(ElasticsearchVersion elasticsearchVersion, IndexInfo indexInfo, SeaTunnelRowType seaTunnelRowType) {
         this.indexTypeSerializer = IndexTypeSerializerFactory.getIndexTypeSerializer(elasticsearchVersion, indexInfo.getType());
         this.indexSerializer = IndexSerializerFactory.getIndexSerializer(indexInfo.getIndex(), seaTunnelRowType);
         this.seaTunnelRowType = seaTunnelRowType;
+        this.keyExtractor = KeyExtractor.createKeyExtractor(seaTunnelRowType, indexInfo.getPrimaryKeys(), indexInfo.getKeyDelimiter());
     }
 
     @Override
     public String serializeRow(SeaTunnelRow row) {
+        switch (row.getRowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                return serializeUpsert(row);
+            case UPDATE_BEFORE:
+            case DELETE:
+                return serializeDelete(row);
+            default:
+                throw new ElasticsearchConnectorException(
+                    CommonErrorCode.UNSUPPORTED_OPERATION, "Unsupported write row kind: " + row.getRowKind());
+        }
+    }
+
+    private String serializeUpsert(SeaTunnelRow row) {
+        String key = keyExtractor.apply(row);
+        Map<String, Object> document = toDocumentMap(row);
+
+        try {
+            if (key != null) {
+                Map<String, String> upsertMetadata = createMetadata(row, key);
+                return String.format("{ \"update\" : %s }\n{ \"doc\" : %s, \"doc_as_upsert\" : true }",
```

Comment:
`serializeUpsert ` will be called by every row. So, I suggest we use better way to process String.